models/sharky/solana/sharky_solana_events.sql

```diff
@@ -0,0 +1,109 @@
+          AND block_time >= date_trunc("day", now() - interval '1 week')
+        {% endif %}
+          AND address = '{{sharky_smart_contract}}'
+    ),
+    sol_price AS (
+        SELECT minute,
+               price
+        FROM {{ source('prices', 'usd') }}
+        WHERE
+        blockchain IS NULL
+        AND symbol = 'SOL'
+        {% if not is_incremental() %}
+        AND minute >= '{{ project_start_date }}'
+        {% endif %}
+        {% if is_incremental() %}
+        AND minute >= date_trunc("day", now() - interval '1 week')
+        {% endif %}
+    ),
+    filtered_txs AS (
+        SELECT *
+        FROM {{ source('solana','transactions') }} tx
+        {% if not is_incremental() %}
+        WHERE tx.block_time >= '{{ project_start_date }}'
+        {% endif %}
+        {% if is_incremental() %}
+        WHERE tx.block_time >= date_trunc("day", now() - interval '1 week')
+        {% endif %}
+    ),
+    events AS (
+        SELECT 'solana'                                                        AS blockchain,
+               'sharky'                                                        AS project,
+               signatures[0]                                                   AS tx_hash,
+               block_date,
+               block_time,
+               CAST(block_slot AS BIGINT)                                      AS block_number,
+               (abs(post_balances[0] - pre_balances[0]) / 1e9) * p.price       AS amount_usd,
+               (abs(post_balances[0] - pre_balances[0]) / 1e9)                 AS amount_original,
+               CAST(abs(post_balances[0] - pre_balances[0]) AS DECIMAL(38, 0)) AS amount_raw,
+               filter(
+                           instructions,
+                           x -> x.executing_account = '{{sharky_smart_contract}}'
+                   )                                                           AS sharky_instructions,
+               CASE
+                   WHEN array_contains(log_messages, 'Program log: Instruction: OfferLoan') THEN 'Offer'
+                   WHEN array_contains(log_messages, 'Program log: Instruction: TakeLoan') THEN 'Take'
+                   WHEN array_contains(log_messages, 'Program log: Instruction: RescindLoan') THEN 'Rescind'
+                   WHEN
+                       (
+                               array_contains(log_messages, 'Program log: Instruction: RepayLoan')
+                               OR array_contains(log_messages, 'Program log: Instruction: RepayLoanEscrow')
+                           ) THEN 'Repay'
+                   WHEN
+                       (
+                               array_contains(log_messages, 'Program log: Instruction: ForecloseLoan')
+                               OR array_contains(log_messages, 'Program log: Instruction: ForecloseLoanEscrow')
+                           ) THEN 'Foreclose'
+                   ELSE 'Other' END                                            AS evt_type,
```

Comment:
Maybe this part can be further optimized, I am not sure.
But string matching is computationally expensive.
If we can calculate `evt_type` in other method(like make use of other columns in tx table), or maybe shorter matching pattern, we may have better performance in querying.